src/accumulator/proof.rs

```diff
@@ -0,0 +1,310 @@
+                    break;
+                }
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+        let mut targets = self.targets.clone();
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes: Vec<_> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(del_hashes.to_owned())
+            .collect();
+
+        let mut row = 0;
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
+
+            if let Ok((start, end)) = res {
+                let mut i = start as usize;
+
+                // Iterates over a row
+                while i <= end as usize {
+                    if nodes.len() == 0 {
+                        return Err(String::from("Malformed tree"));
+                    }
+
+                    // If the current position is a root, we add that to our result and don't go any further
+                    if util::is_root_position(nodes[i].0, stump.leafs, total_rows) {
+                        calculated_root_hashes.push(nodes[i].1);
+                    } else if nodes.len() > i + 1 && (nodes[i].0 | 1 == nodes[i + 1].0) {
+                        // If the next node is the current node's sibling, we hash both nodes together
+
+                        let hash = types::parent_hash(&nodes[i].1, &nodes[i + 1].1);
+                        let next_to_prove = util::parent(nodes[i].0, total_rows);
+
+                        targets.push(next_to_prove);
+                        targets.sort();
+
+                        Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+
+                        i += 1;
+                    } else {
+                        if nodes.len() == 0 || self.hashes.len() == 0 {
+                            return Err(String::from("MissingNode"));
+                        }
+
+                        // If the next node is not my sibling, the hash must be passed inside the proof
+                        let hash = if util::is_left_niece(nodes[i].0) {
+                            types::parent_hash(&nodes[i].1, &self.hashes[proof_hashes_index])
+                        } else {
+                            types::parent_hash(&self.hashes[proof_hashes_index], &nodes[i].1)
```

Comment:
`proof_hashes_index` is another Go hack since this is the best way to do iteration in Go. With Rust, you have `.iter()` on vectors so it'd be better have that instead of doing incrementing.

In the beginning of `create_root_hashes` function, we can do something like:

`let proof_hashes_iter = self.hashes.iter()`

and call `proof_hashes_iter.next()` in this branch. If the result is `None`, return an error.